src/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -157,6 +157,127 @@ public TResult TryGetLast(out bool found)
             }
         }
 
+        private sealed partial class SelectRangeIterator<TResult> : Iterator<TResult>, IPartition<TResult>
+        {
+            private readonly int _start;
+            private readonly int _end;
+            private readonly Func<int, TResult> _selector;
+
+            public SelectRangeIterator(int start, int end, Func<int, TResult> selector)
+            {
+                Debug.Assert(start < end);
+                Debug.Assert((end - start) <= int.MaxValue);
+                Debug.Assert(selector != null);
+
+                _start = start;
+                _end = end;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone() =>
+                new SelectRangeIterator<TResult>(_start, _end, _selector);
+
+            public override bool MoveNext()
+            {
+                if (_state < 1 || _state == (_end - _start + 1))
+                {
+                    Dispose();
+                    return false;
+                }
+
+                int index = _state++ - 1;
+                _current = _selector(_start + index);
+                return true;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector) =>
```

Comment:
nit: Is there a more descriptive name that we can use instead of `TResult2`?

Reply:
`TTo` ?

Reply:
Ditto